akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpResponseParser.scala

```diff
@@ -154,41 +154,51 @@ private[http] class HttpResponseParser(protected val settings: ParserSettings, p
           startNewMessage(input, bodyStart)
       }
 
-    if (statusCode.allowsEntity && (contextForCurrentResponse.get.requestMethod != HttpMethods.HEAD)) {
-      teh match {
-        case None ⇒ clh match {
-          case Some(`Content-Length`(contentLength)) ⇒
-            if (contentLength == 0) finishEmptyResponse()
-            else if (contentLength <= input.size - bodyStart) {
-              val cl = contentLength.toInt
-              emitResponseStart(strictEntity(cth, input, bodyStart, cl))
-              setCompletionHandling(HttpMessageParser.CompletionOk)
-              emit(MessageEnd)
-              startNewMessage(input, bodyStart + cl)
-            } else {
-              emitResponseStart(defaultEntity(cth, contentLength))
-              parseFixedLengthBody(contentLength, closeAfterResponseCompletion)(input, bodyStart)
-            }
-          case None ⇒
-            emitResponseStart {
-              StreamedEntityCreator { entityParts ⇒
-                val data = entityParts.collect { case EntityPart(bytes) ⇒ bytes }
-                HttpEntity.CloseDelimited(contentType(cth), HttpEntity.limitableByteSource(data))
-              }
-            }
+    if (statusCode.allowsEntity) {
+      contextForCurrentResponse.get.requestMethod match {
+        case HttpMethods.HEAD ⇒ clh match {
+          case Some(`Content-Length`(contentLength)) if contentLength > 0 ⇒
+            emitResponseStart(defaultEntity(cth, contentLength))
```

Comment:
I think we need something extra here to create a `StrictEntityCreator` with a Default entity that has a `Source.empty`. That's because `StrictEntityCreator` will later correctly cancel the empty stream it gets created with.